### PR TITLE
Fix df.to_csv() for string arrays when encoded in utf-8

### LINKED ISSUE #10813: to_csv with lists of strings and unicode encoding produces wrong output
If I have a dataframe with cells containing lists of strings (or unicode strings), then these lists are broken when I use `to_csv()` with the `encoding` parameter set. The error does not occur if the `encoding` is not set.

Here is an example (using pandas version 0.16.2):

```
df = pd.DataFrame.from_records(
    [('Mary S.',['Detroit, MI','New York, NY']),
     ('John U.',[u'Atlanta, GA',u'Paris, France'])],
    columns=['name','residences'])
df.to_csv('ascii.csv')
df.to_csv('utf8.csv',encoding='utf-8')
```

The ascii-encoded CSV file is fine. (contents of 'ascii.csv' below)

```
,name,residences
0,Mary S.,"['Detroit, MI', 'New York, NY']"
1,John U.,"[u'Atlanta, GA', u'Paris, France']"
```

But the unicode CSV file fails to quote the strings within the lists. (contents of 'utf8.csv' below)

```
,name,residences
0,Mary S.,"[Detroit, MI, New York, NY]"
1,John U.,"[Atlanta, GA, Paris, France]"
```

This results in the data being impossible to recover. For example, if I load this file using `read_csv()`, the relevant cells are treated as strings, and cannot be accurately recast as lists.

The behavior is the same using `encoding='utf-16'` but I didn't check any other encodings.

### PR DESCRIPTION
**Pull request by rtkaleta**
So it looks like `df.to_csv()` is now working correctly for string arrays when using the `ascii` encoding but it is still broken when using `utf-8`.

### CODE REVIEW

File: pandas/tests/io/formats/test_to_csv.py [hunk 1]
```diff
@@ -223,3 +223,27 @@ def test_to_csv_multi_index(self):
 
         exp = "foo\nbar\n1\n"
         assert df.to_csv(index=False) == exp
+
+    def test_to_csv_string_array(self):
+        # GH 10813
+        str_array = [{'names': ['foo', 'bar']}, {'names': ['baz', 'qux']}]
+        df = pd.DataFrame(str_array)
+        expected_ascii = '''\
+,names
```
Comment (jreback) [inline pandas/tests/io/formats/test_to_csv.py]: so if you make this 2 test functions, then you can xfail the non-working one (to at least get things passing)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by rtkaleta** [comment]
@jreback It seems the current behaviour stems from the fact that pandas' own `UnicodeWriter` calls `pprint_thing` without `quote_strings=True` so then:

```python
>>> from pandas.io.formats.printing import pprint_thing
>>> pprint_thing([u'foo', u'bar'])
u'[foo, bar]'
```
instead of the more intuitive (at least to me):
```python
>>> pprint_thing([u'foo', u'bar'], quote_strings=True)
u"[u'foo', u'bar']"
```

A couple of questions come to my mind:
1) Why do we have our own `UnicodeWriter` [here](https://github.com/pandas-dev/pandas/blob/master/pandas/io/formats/format.py#L1642) instead of [unicodecsv.writer](https://github.com/jdunck/python-unicodecsv)?
2) Better to expose the `quote_strings` (or similar) parameter to the `to_csv` caller, or is the behaviour I expect so ubiquitously intuitive that we should be changing things under the hood?

**Comment by jreback** [comment]
will have a look

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @rtkaleta 

love for you to take a stab at fixing the xfailed unicode case!

**Pull request merged**